DMTN-148.tex

```diff
@@ -50,9 +50,237 @@
+the issue.
+
+
+\section{The End State}
+
+The final state that we wish to get to has the following properties:
+
+\begin{itemize}
+\item All calibration products are either lsst.afw.image.Exposure
+  image data or an explicit calibration class in the \verb|ip_isr| package.
+\item All calibration products are constructed from a DM Task in the
+  \verb|cp_pipe| package, and have the input data and configuration
+  parameters persisted with the calibration itself.  This guarantees
+  that any product can be reconstructed in the future.
+\item These provenance, config, and final product files may be saved
+  to an \verb|obs_CAMERA_data| package on git for convenience.  This also allows
+  these files to be migrated to other Butler repositories in a
+  consistent fashion.
+\item All calibration products are accepted for use and have a date
+  range set manually, preventing test runs from being used
+  inadvertently.  Calibrations ingested from git should supply
+  appropriate date ranges.
+\item Calibration information stored in a lsst.afw.cameraGeom object
+  may be used, but is always assumed to be overridden by supplied
+  calibration products.  No new calibration information should be
+  stored in these objects, and currently existing calibration
+  information should be deprecated.
+
+\end{itemize}
+
+\section{The Current State}
+
+\subsection{Standard Calibrations}
+
+The classic calibration products, \verb|BIAS|, \verb|DARK|, \verb|FLAT|, and \verb|FRINGE|, all
+write FITS files that are indexed in Butler's calibRegistry.sqlite3
+repository database.  These files are created by
+\verb|pipe_drivers|/constructCalibs.py, and are ingested into the
+repository upon acceptance by \verb|pipe_tasks|/ingestCalibs.py.  This
+ingestion process sets the 'validStart' and 'validEnd' values to
+restrict the usage of the calibration, as well as the 'calibDate',
+which is used (along with calibration type, filter, and detector) to
+uniquely identify the calibration on disk.
+
+Although not formally a "classic" calibration product, the \verb|SKY|
+calibration is produced, stored, and accessed the same way.
+
+These calibrations are required to have the following header keywords
+defined, which are used during ingest to define the rows of the
+appropriate calibration database table.  I believe only \verb|OBSTYPE|
+is required right now.
+
+\begin{tabular}{l l}
+  Keyword & Description \\
+  \hline
+  \verb|OBSTYPE| & Calibration type contained in this file. \\
+  \verb|INSTRUME| & Instrument/camera this calibration belongs to. \\
+  \verb|DETECTOR| & Integer identifier for the ccd/detector. \\
+  \verb|FILTER| & Name of the physical filter or ``NONE''. \\
+  \verb|CALIB_CREATION_DATE| & Creation date of the calibration. \\
+  \verb|CALIB_CREATION_TIME| & Creation time of the calibration. \\
+  \verb|CALIBDATE| & Concatenation of \verb|CALIB_CREATION_DATE| and \verb|CALIB_CREATION_TIME|.  Unique for set of OIDF.\\
```

Comment:
Sorry.  Reading this here after commenting above.  I think it's still necessary to think about the "reprocess everything at once" use case.

Reply:
I've removed that, and agree we should be using something better than a date string.

I also didn't understand that QE curves are amp-by-amp.  Are they persisted this way as well?  I don't think I know how they're used, as that's somewhere other than ip_isr.

Reply:
Good point.  The QE curves are per amp, but they are stored in a single file per detector, so I guess we are safe.  Sorry for the false alarm.